### PR TITLE
Ports SSmetrics from bee, which was ported from Para

### PR DESCRIPTION
**Pull request by MarkSuckerberg**
## About The Pull Request
Ports: beestation/beestation-hornet#5152 which was a port of ParadiseSS13/Paradise#16549, as well as ports BeeStation/BeeStation-Hornet#8007.

Adds a subsystem that optionally shoots server metrics to an elasticsearch server. I think you can guess why I decided to port this.

## Why It's Good For The Game
Good for monitoring the server's status, most importantly the specific subsystems so that we can tell if something is going awry.

## Changelog

:cl: AffectedArc07, Crossedfall
add: Added a subsystem which can send server metrics straight into ElasticSearch.
/:cl:

### CODE REVIEW

File: code/controllers/subsystem/metrics.dm [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+	Not only is changing this a very bad idea due to elasticsearch being very finnicky with data formatting,
+	but if you edit this subsystem and its fields, you invalidate a lot of existing data
+
+	Dont touch this shit without speaking to crossed or AA07 first
+
+*/
+SUBSYSTEM_DEF(metrics)
+	name = "Metrics"
+	wait = 30 SECONDS
+	runlevels = RUNLEVEL_LOBBY | RUNLEVEL_SETUP | RUNLEVEL_GAME | RUNLEVEL_POSTGAME // ALL THE LEVELS
```
Comment (ZephyrTFA) [inline code/controllers/subsystem/metrics.dm]:
```suggestion
	runlevels = ALL
```

File: code/controllers/subsystem/metrics.dm [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+	flags = SS_KEEP_TIMING // This needs to ingest every 30 IRL seconds, not ingame seconds.
+	/// The real time of day the server started. Used to calculate time drift
+	var/world_init_time = 0 // Not set in here. Set in world/New()
+
+/datum/controller/subsystem/metrics/Initialize(start_timeofday)
+	if(!CONFIG_GET(flag/elasticsearch_metrics_enabled))
+		flags |= SS_NO_FIRE // Disable firing to save CPU
+	return ..()
```
Comment (ZephyrTFA) [inline code/controllers/subsystem/metrics.dm:24]:
```suggestion
	..()
```

File: code/controllers/subsystem/metrics.dm [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
```
Comment (ZephyrTFA) [inline code/controllers/subsystem/metrics.dm:25]:
```suggestion
```

File: code/controllers/subsystem/metrics.dm [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+*/
+SUBSYSTEM_DEF(metrics)
+	name = "Metrics"
+	wait = 30 SECONDS
+	runlevels = RUNLEVEL_LOBBY | RUNLEVEL_SETUP | RUNLEVEL_GAME | RUNLEVEL_POSTGAME // ALL THE LEVELS
+	flags = SS_KEEP_TIMING // This needs to ingest every 30 IRL seconds, not ingame seconds.
+	/// The real time of day the server started. Used to calculate time drift
+	var/world_init_time = 0 // Not set in here. Set in world/New()
+
```
Comment (ZephyrTFA) [inline code/controllers/subsystem/metrics.dm:19]:
```suggestion

GENERAL_DATUM_PROTECT(/datum/controller/subsystem/metrics)

```

File: code/controllers/subsystem/garbage.dm [hunk 5]
```diff
@@ -56,6 +56,19 @@ SUBSYSTEM_DEF(garbage)
...
+	// You can calculate TGCR in kibana
+	cust["total_harddels"] = totaldels
+	cust["total_softdels"] = totalgcs
+	var/i = 0
+	for(var/list/L in queues)
+		i++
+		cust["queue_[i]"] = length(L)
+
+	.["custom"] = cust
+
```
Comment (ZephyrTFA) [inline code/controllers/subsystem/garbage.dm:71]:
```suggestion
```

**Review by ZephyrTFA — CHANGES_REQUESTED** [review]

File: code/controllers/subsystem/metrics.dm (hunk 2 above)
Comment (MarkSuckerberg) [reply code/controllers/subsystem/metrics.dm:24]: You sure?

**Review by ZephyrTFA — APPROVED** [review]

**Pull request merged**